pybop/builders/python.py

```diff
@@ -1,53 +1,127 @@
-import pybop
+from typing import Callable
+
 from pybop import PythonProblem
 from pybop.builders.base import BaseBuilder
-from pybop.problems.base_problem import Problem
 
 
 class Python(BaseBuilder):
     """
-    dataset : pybop.Dataset or dict, optional
-        The dataset to be used in the simulation construction.
+    Builder for Python-based problems.
+
+    This builder creates problems using custom Python functions instead of
+    specialised simulation frameworks. It supports both standard models and
+    models with sensitivity analysis.
+
+    If this problem is used by Pybop Optimisation or Sampling classes, the
+    functions will be minimised.
+
+    Examples
+    --------
+    >>> builder = Python()
+    >>> builder.add_func(my_model_function, weight=1.5)
+    >>> problem = builder.build()
     """
 
     def __init__(self):
-        self._model = None
-        self._params = None
-        self._costs = []
-        self._cost_weights = []
-        self._func = None
+        super().__init__()
+        self._models: list[Callable] = []
+        self._models_with_sens: list[Callable] = []
+        self._weights: list[float] = []
+
+    def add_func(self, model: Callable, weight: float = 1.0) -> "Python":
+        """
+        Add a simulation function to the problem.
+
+        Parameters
+        ----------
+        model : Callable
+            Function with signature: func(params: Union[List[float], float]) -> dict
+            Returns dict where keys match signal names and values are predictions.
+        weight : float, default=1.0
+            Weight for this model in multi-objective optimisation.
 
-    def add_func(self, func):
-        self._func = func
+        Returns
+        -------
+        Python
+            Self for method chaining.
 
-    def add_cost(self, cost: pybop.BaseCost, weight: float = 1.0) -> None:
+        Raises
+        ------
+        TypeError
+            If model is not callable.
         """
-        Adds cost to the problem.
-        ToDo: BaseCost needs to be updated to support array/list-like execution
+        if not callable(model):
+            raise TypeError("Model must be callable")
+
+        self._models.append(model)
+        self._weights.append(weight)
+        return self
+
+    def add_func_with_sens(
+        self, model_with_sens: Callable, weight: float = 1.0
+    ) -> "Python":
         """
-        self._costs.append(cost)
-        self._cost_weights.append(weight)
+        Add a simulation function with sensitivity analysis to the problem.
 
-    def build(self) -> Problem:
-        """Build a pure python problem."""
+        Parameters
+        ----------
+        model_with_sens : Callable
+            Function with signature: func(params: Union[List[float], float]) -> Tuple[dict, dict]
```

Comment:
can I check what the signature of the function should be? When you use it in the python problem, it looks like the signature should be:
```func(params:  np.ndarray) -> float```